File: 2594_Minimum_Time_to_Repair_Cars/Solution.py
from typing import List
from math import sqrt


class Solution:
    def canRepair(self, x, ranks, cars):
        total = 0
        for rank in ranks:
            total += int(sqrt(x // rank))
            if total >= cars:
                return True
        return False

    def repairCars(self, ranks: List[int], cars: int) -> int:
        ranks = sorted(ranks)
        low = 1
        high = ranks[0] * cars * cars
        answer = high
        while (low <= high):
            mid = low + (high - low) // 2
            if self.canRepair(mid, ranks, cars):
                answer = mid
                high = mid - 1
            else:
                low = mid + 1
        return answer

File: 2594_Minimum_Time_to_Repair_Cars/Solution.java
import java.util.*;
class Solution {
    public boolean canRepair(long x, int[] ranks, int cars){
        int total = 0;
        for(int rank: ranks){
            total += Math.sqrt(x/rank);
            if(total>=cars){
                return true;
            }
        }
        return false;
    }
    public long repairCars(int[] ranks, int cars) {
        Arrays.sort(ranks);
        long low = 1;
        long high = (long) ranks[0]*cars*cars;
        long answer = high;
        while(low<=high){
            long mid = low+ (high-low)/2;
            if(canRepair(mid, ranks, cars)){
                answer = mid;
                high = mid-1;
            }
            else{
                low = mid+1;
            }
        }
        return answer;

    }
}